CMakeLists.txt

```diff
@@ -301,7 +301,7 @@ include_directories(${PROJECT_SOURCE_DIR}/src/lib)
 include_directories(${PROJECT_SOURCE_DIR}/src/lib/small/include)
 include_directories(${PROJECT_SOURCE_DIR}/src/lib/small/third_party)
 include_directories(${PROJECT_SOURCE_DIR}/src/lib/core)
-include_directories(${PROJECT_SOURCE_DIR}/third_party)
+include_directories(SYSTEM ${PROJECT_SOURCE_DIR}/third_party)
```

Comment:
Setting the `SYSTEM` option may have some extra side-effects aside ignoring warnings. According to the [doc](https://cmake.org/cmake/help/latest/command/include_directories.html), "these fixed-install system files not being considered in dependency calculations". I don't think we should set it for our internal files that may actually be updated. @Totktonada should know better.

Reply:
@locker you quoted out of context:

> Signalling this setting might achieve effects such as the compiler skipping warnings, or these fixed-install system files not being considered in dependency calculations - see compiler docs.

The key words here are _see compiler docs_. AFAI, what CMake does for [**GNU**](https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html) is passing the the `-isystem` flag:
> Header files found in directories added to the search path with the -isystem and -idirafter command-line options are treated as system headers for the purposes of diagnostics.

Same goes for [**Clang**](https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-in-system-headers)

Reply:
> @Totktonada should know better.

Hm. Someone called me a CMake professional? I'm definitely not one.

To be honest, personally I think that CMake is even worse than Bourne shell, but taking into account that it was born 20 years afterwards, it is purely insane thing. And it has worst documentation I ever saw in a popular project.

----

Vladimir said about cmake dependency resolution (rebuild everything that needs to be rebuilt after changing a file). It is a cmake's thing, not a compiler's one.

I'm a bit worried about that too, but not only about that.

There are libraries/headers in the `third_party` directory that are copied verbatim, there are ones that are patched at some points, there are ones that were rewritten significantly. There are ones with evolving upstreams and time-to-time pulldowns into our source tree, but there are ones, where upstreams changing rarely.

IOW, there are libraries of different kinds and it looks improper to disable warning in all of them due to one or two affected ones.

I see the following instruments around:

* Fix warnings. In context of libev it seems counter-productive, because it makes pulldowns more cumbersome.
* Use the `include_directories(SYSTEM <...>)` hack.
* Use a [precompiled headers hack](https://gitlab.kitware.com/cmake/cmake/-/issues/21219).
* Add pragmas for pushing/popping the diagnostics for specific headers that needs it.

The latter seems less intruisive as in affecting other third party libraries as well as in changing library's code itself.

Reply:
Thanks for elaborating on the problem and for the suggested solutions: I chose to go with the latter.

Reply:
Looks like ignoring `-Wstrict-aliasing` won't work at all: https://godbolt.org/z/Gzx1MG7Wh — I will try to report this bug on GCC.

Reply:
Yes, but the warning is supposed to be issued at the point where macro is defined — you can try out other warnings, and they will not be issued.

Reply:
> especially since old GNU compilers don't seem to support the system_header pragma well

Looks like there was also a bug in **GCC** with system headers that was only fixed in GCC 8: https://godbolt.org/z/xr54exzn7.> especially since old GNU compilers don't seem to support the system_header pragma well

Looks like there was also a bug in **GCC** with system headers that was only fixed in GCC 8: https://godbolt.org/z/xr54exzn7.

Reply:
Warnings are for developers and developers are usually use relatively fresh compiler version. I see no problem in disabling a particular warning for the whole project for an old (but still used) compiler version. It is the same as we previously do for `-Wubused-value` and `-Wunknown-pragmas`.

I would even say that we can ignore the warnings, but we have `-Werror` builds and it would be inconvenient.

Reply:
I suggest the following solution:

* use `system_header` pragma;
* for GNU compilers older than 8 use `-Wno-strict-aliasing`.

Reply:
> I would prefer to disable only necessary warnings, because I would like to at least look at new warnings if they'll appear.

As I pointed out in the commit message, libev is a special case, where the practice is to simply ignore all warnings coming from it.

Reply:
> I suggest the following solution:
> 
> * use `system_header` pragma for libev;
> * leave `-Wno-strict-aliasing` for GNU compilers older than 8.

Agreed.

> As I pointed out in the commit message, libev is a special case, for which the practice advocated even by library authors themselves is to simply ignore all compiler warnings coming from it.

I would still prefer to analyze new warnings explicitly disregarding on how self-confident authors of third party libraries are.

But since we have no working way to selectively disable this particular warning when it comes from a particular header, I'm okay with the proposed solution.